From 6c7f6ef6e8b6e61a97be029d98e20922a614edf0 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Fri, 3 Jul 2020 12:37:37 +0800
Subject: [PATCH 36/98] backend-drm: Support controlling compositor dynamically

Use config file to control compositor's state.

Default config file is "/tmp/.weston_drm.conf", can override with
"WESTON_DRM_CONFIG" environment.

Supported configs format is "compositor:<key>:<value>", for
example:
echo "compositor:state:sleep" > /tmp/.weston_drm.conf
echo "compositor:state:block" > /tmp/.weston_drm.conf
echo "compositor:state:freeze" > /tmp/.weston_drm.conf
echo "compositor:state:offscreen" > /tmp/.weston_drm.conf
echo "compositor:state:off" > /tmp/.weston_drm.conf
echo "compositor:state:on" > /tmp/.weston_drm.conf
echo "compositor:hotplug:off" > /tmp/.weston_drm.conf
echo "compositor:hotplug:on" > /tmp/.weston_drm.conf
echo "compositor:hotplug:force" > /tmp/.weston_drm.conf
echo "compositor:cursor:hide" > /tmp/.weston_drm.conf
echo "compositor:cursor:show" > /tmp/.weston_drm.conf

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 libweston/backend-drm/drm.c | 55 ++++++++++++++++++++++++++++++++++++-
 libweston/compositor.c      |  7 ++++-
 2 files changed, 60 insertions(+), 2 deletions(-)

diff --git a/libweston/backend-drm/drm.c b/libweston/backend-drm/drm.c
index 2cb4a7e..2f3d27c 100644
--- a/libweston/backend-drm/drm.c
+++ b/libweston/backend-drm/drm.c
@@ -4392,6 +4392,55 @@ config_handle_output(struct drm_backend *b, const char *name,
 	}
 }
 
+static void
+config_handle_compositor(struct drm_backend *b, const char *key,
+			 const char *value)
+{
+	if (!strncmp(key, "state", strlen("state"))) {
+		if (!strncmp(value, "sleep", strlen("sleep"))) {
+			weston_compositor_sleep(b->compositor);
+		} else if (!strncmp(value, "block", strlen("block"))) {
+			udev_input_disable(&b->input);
+		} else if (!strncmp(value, "freeze", strlen("freeze"))) {
+			udev_input_disable(&b->input);
+			weston_compositor_offscreen(b->compositor);
+		} else if (!strncmp(value, "offscreen", strlen("offscreen"))) {
+			/* HACK: offscreen + DPMS off */
+			weston_compositor_sleep(b->compositor);
+			weston_compositor_offscreen(b->compositor);
+			udev_input_enable(&b->input);
+		} else if (!strncmp(value, "off", strlen("off"))) {
+			udev_input_disable(&b->input);
+			weston_compositor_sleep(b->compositor);
+		} else {
+			/* HACK: Force waking from offscreen */
+			if (b->compositor->state == WESTON_COMPOSITOR_OFFSCREEN)
+				b->compositor->state = WESTON_COMPOSITOR_IDLE;
+
+			weston_compositor_wake(b->compositor);
+			weston_compositor_damage_all(b->compositor);
+
+			if (b->input.suspended)
+				udev_input_enable(&b->input);
+		}
+	} else if (!strncmp(key, "hotplug", strlen("hotplug"))) {
+		if (!strncmp(value, "off", strlen("off")))
+			wl_event_source_fd_update(b->udev_drm_source, 0);
+		else if (!strncmp(value, "on", strlen("on")))
+			wl_event_source_fd_update(b->udev_drm_source,
+						  WL_EVENT_READABLE);
+		else if (!strncmp(value, "force", strlen("force")))
+			hotplug_timer_handler(b);
+	} else if (!strncmp(key, "cursor", strlen("cursor"))) {
+		if (!strncmp(value, "hide", strlen("hide")))
+			b->compositor->hide_cursor = true;
+		else if (!strncmp(value, "show", strlen("show")))
+			b->compositor->hide_cursor = false;
+
+		weston_compositor_damage_all(b->compositor);
+	}
+}
+
 static int
 config_timer_handler(void *data)
 {
@@ -4423,7 +4472,9 @@ config_timer_handler(void *data)
 
 	/**
 	 * Parse configs, formated with <type>:<key>:<value>
-	 * For example: "output:all:rotate90"
+	 * For example:
+	 *	output:all:rotate90
+	 *	compositor:state:off
 	 */
 	while (3 == fscanf(conf_fp,
 			   "%" STR(MAX_CONF_LEN) "[^:]:"
@@ -4431,6 +4482,8 @@ config_timer_handler(void *data)
 			   "%" STR(MAX_CONF_LEN) "[^\n]%*c", type, key, value)) {
 		if (!strcmp(type, "output"))
 			config_handle_output(b, key, value);
+		else if (!strcmp(type, "compositor"))
+			config_handle_compositor(b, key, value);
 	}
 
 	fclose(conf_fp);
diff --git a/libweston/compositor.c b/libweston/compositor.c
index b0eba75..16dae0f 100644
--- a/libweston/compositor.c
+++ b/libweston/compositor.c
@@ -5353,6 +5353,10 @@ weston_compositor_wake(struct weston_compositor *compositor)
 {
 	uint32_t old_state = compositor->state;
 
+	/* HACK: Avoid waking up by input */
+	if (old_state == WESTON_COMPOSITOR_OFFSCREEN)
+		return;
+
 	/* The state needs to be changed before emitting the wake
 	 * signal because that may try to schedule a repaint which
 	 * will not work if the compositor is still sleeping */
@@ -5360,10 +5364,11 @@ weston_compositor_wake(struct weston_compositor *compositor)
 
 	switch (old_state) {
 	case WESTON_COMPOSITOR_SLEEPING:
+		wl_signal_emit(&compositor->wake_signal, compositor);
+		/* fall through */
 	case WESTON_COMPOSITOR_IDLE:
 	case WESTON_COMPOSITOR_OFFSCREEN:
 		weston_compositor_dpms(compositor, WESTON_DPMS_ON);
-		wl_signal_emit(&compositor->wake_signal, compositor);
 		/* fall through */
 	default:
 		wl_event_source_timer_update(compositor->idle_source,
-- 
2.20.1

